[JENKINS-28335] - Git username and password binding (GSoC 2021) (#1104)
jenkinsci/git-plugin: README.adoc

image:images/multibranch-pipeline.png[link=https://youtu.be/B_2FXWI6CWg]

[#credential-binding]
=== Git Credentials Binding

> The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *SSH* protocols using command line git in a Pipeline job.

Reviewer: SSH support is not yet ready.  Let's not mention it yet.
```suggestion
The git plugin provides `Git Username and Password` binding that allows authenticated git operations over *HTTP* and *HTTPS* protocols using command line git in a Pipeline job.
```

> The git credential bindings are accessible through the `withCredentials` step of the link:https://plugins.jenkins.io/credentials-binding/[Credentials Binding] plugin.

Reviewer: Hyperlink to the step documentation?
```suggestion
The git credential bindings are accessible through the link:https://www.jenkins.io/doc/pipeline/steps/credentials-binding/#withcredentials-bind-credentials-to-variables[`withCredentials`] step of the link:https://plugins.jenkins.io/credentials-binding/[Credentials Binding] plugin.
```

The binding retrieves credentials from the link:https://plugins.jenkins.io/credentials/[Credentials] plugin.

==== Git Username and Password Binding

This binding provides authentication support over *HTTP* protocol using command line git in a Pipeline job.

Procedure::

. Click the Pipeline Syntax _Snippet Generator_ and choose the `withCredentials` step, add Git Username and Password binding.
. Choose the required credentials and Git tool name, specific to the generated Pipeline snippet.

image:images/git-credentials-usernamepassword-binding-pipline-job.png[Git-Username-and Password-Binding-Pipeline-Job]

> NOTE: By default, two variable bindings, `GIT_USERNAME` and `GIT_PASSWORD`, bind with user’s credential i.e. username and password respectively, in a Pipeline jobs.

Reviewer: This seems like an interesting implementation detail, but placing it in a NOTE gives it more emphasis that I think is justified.  The user should not need to do anything with either of these variables.  They are internal implementation details.
```suggestion
Two variable bindings are used, `GIT_USERNAME` and `GIT_PASSWORD`, to pass the username and password to `sh`, `bat`, and `powershell` steps inside the `withCredentials` block of a Pipeline job.
```

  ViewGit project name.
  For example, `my-project`

[#git-bindings]
== Git Bindings

The git plugin provides one binding to support authenticated git opreations over *HTTP* or * HTTPS* protocol, namely `Git Username and Password`.
The git plugin depeneds on the Credential Binding Plugin to support these bindings.

To access the `Git Username and Password` binding in a pipeline job, visit <<credential-binding>>

> Incase of a FreeStyle project the steps are as follows:

Reviewer: Remove the 'Incase' word
```suggestion
Freestyle projects can use git credential binding with the following steps:
```

. Check the box _Use secret text(s) or file(s)_, add Git Username and Password binding.

. Choose the required credentials and Git tool name.

image:images/git-credentials-usernamepassword-binding-freestyle-project.png[Git-Username-and Password-Binding-Freestyle-project]

> NOTE: By default, two variable bindings, `GIT_USERNAME` and `GIT_PASSWORD`, bind with user’s credential i.e. username and password respectively, in a Pipeline jobs.

Reviewer: Remove the NOTE: emphasis here as well
```suggestion
Two variable bindings are used, `GIT_USERNAME` and `GIT_PASSWORD`, to pass the username and password to shell, batch, and powershell steps in a Freestyle job.
```
